### ISSUE TITLE
CorsPolicyService not being called

### ISSUE THREAD

**Issue by thoraj** [IdentityServer/IdentityServer4#3780]
I'm trying to use a custom policyservice, but it is never being called: 

Our experience is similar to what is described in a previous issue: https://github.com/IdentityServer/IdentityServer4/issues/3075

The implementation of ICorsPolicyService is simple:
```    
public class VerjiCorsPolicyService : ICorsPolicyService
    {
        private readonly string[] _corsOrigins;

        public VerjiCorsPolicyService(string[] corsOrigins)
        {
            _corsOrigins = corsOrigins.Select(o => o.Replace("http://", "").Replace("*.", "")).Distinct().ToArray();
        }

        public async Task<bool> IsOriginAllowedAsync(string origin)
        {
            if (_corsOrigins.Any(o => o.EndsWith(origin)))
            {
                return true;
            }

            return false;

        }
    }
```

I've made sure to register the service _before_ IdentityServer: 
```
            var verjiCorsService= new VerjiCorsPolicyService(corsOrigins);
            services.AddSingleton<ICorsPolicyService>(verjiCorsService);
```
The log says:
```
31 Oct 2019 10:42:48.295
 CorsPolicyService did not allow origin: https://admros.staging.wc.verji.io
31 Oct 2019 10:42:48.295
 Origin https://admros.staging.wc.verji.io is allowed: false
31 Oct 2019 10:42:48.292
 CORS request made for path: /.well-known/openid-configuration from origin: https://admros.staging.wc.verji.io
```
But `IsOriginAllowedAsync`is never called. 

Is there some other way we should register the custom policy service to make it take effect?

**Comment by brockallen** [comment]
Register it after - I don't recall if we use TryAdd or not for our default.

**Comment by thoraj** [comment]
Thanks for your reply, 

I was able to make it work by registering the service _after_ IdentityServer as you suggested. 

But only after removing the standard Asp.Net Core CORS configuration (AddCors+AddPolicy/UseCors).

So something is not right with either IdentityServer or the docs (or both). According to the docs in the section **Mixing IdentityServer’s CORS policy with ASP.NET Core’s CORS policies**, it should be possible to have both standard AspNet Core CORS alongside a custom ICorsPolicyService if you register the custom service **prior** to IdentityServer. We could not make that work.

Since I am only hosting IdentityServer in this service I don't really need to mix the two, so I will continue using our custom implementation of ICorsPolicyService.

(Strangely, using only the standard AspNet Core CORS setup seemed to work (I could log in), but I kept getting those messages in the logs that CORS failed.)

Now with the custom cors service, the log messages are normal.

Thanks, 

Thor A. Johansen
Rosberg AS
